Auth API changes - Async only and add challenge behavior

HaoK opened the pull request:
Changes needed for Authentication changes in security:
- IAuthenticationHandler is now mostly async, AuthenticationManager.Authencate is a sync wrapper
- Http.Abstractions now depends on Http.Features to reuse AuthenticateContext in AuthManager
- Nuke AuthenticateResult, with Authenticate api changed to use AuthenicateContext
- Add sugar methods which just take scheme
- Remove 401 setting behavior for challenge in manager (Handler will be responsible for everything)
- Add ChallengeBehavior enum/property to ChallengeContext
- Add new Forbidden API which maps to (Challenge with Forbidden(403) behavior)

cc @lodejard @davidfowl @Tratcher @divega

Tratcher commented on src/Microsoft.AspNet.Http.Abstractions/project.json:
```diff
@@ -2,6 +2,7 @@
   "version": "1.0.0-*",
   "description": "ASP.NET 5 HTTP object model. HttpContext and family.",
   "dependencies": {
+    "Microsoft.AspNet.Http.Features": "1.0.0-*",
```
@kichalla Didn't you already add this?

kichalla replied:
@Tratcher : I did not check this in actually...i wanted to combine Session related changes with the checkin...

PinpointTownes commented on src/Microsoft.AspNet.Http.Abstractions/Authentication/AuthenticationManager.cs:
```diff
@@ -11,9 +12,24 @@ public abstract class AuthenticationManager
+        public abstract void Authenticate(AuthenticateContext context);
 
-        public abstract Task<AuthenticationResult> AuthenticateAsync(string authenticationScheme);
+        public abstract Task AuthenticateAsync(AuthenticateContext context);
```
Retrieving the `AuthenticationProperties` with this new method will be really cumbersome...

HaoK replied:
That is fairly uncommon which is why we are leaning this way to start (Identity will be using the properties, so when I get there, if its super bad, don't worry I'll definitely add sugar :)

But most of the time, Authenticate() giving you back a claimsprincipal really all you want, without having to dig through the old Result tuple...

PinpointTownes replied:
I'm not sure it's that uncommon: I've been asked at least 3 or 4 times on JabbR how to retrieve the expiration date of a cookie or a token.

HaoK replied:
yeah but its not that hard...

```
var context = new AuthenticateContext(scheme);
httpContext.Authentication.Authenticate(context);
var properties = context.Properties;
```

PinpointTownes replied:
I agree, but it's not really intuitive. And you now have to ensure `Accepted` is true:

```csharp
var context = new AuthenticateContext(scheme);
httpContext.Authentication.Authenticate(context);
if (context.Accepted) {
    var principal = context.Principal;
    ...
}
```

... which is not really simpler than the old API:

```csharp
var principal = (await context.Authentication.AuthenticateAsync(scheme))?.Principal;
if (principal != null) {
    ...
}
```

HaoK replied:
Nah, you don't have to ensure accepted is true, the AuthenticationManager.Authenticate ensures that, or it throws.  That's an implementation detail, I can possibly make that internal

PinpointTownes replied:
Ah yeah, you just have to ensure that the principal is not null, just like the old approach.

HaoK replied:
Yep, the motivation for this change was to make it easier to access the principal, with the intentional tradeoff that accessing Properties/Description is harder.

darrelmiller commented:
Would it be possible for someone to explain the reasoning behind changing to this style of API?  Personally I find the new style a major regression from the point of view of clarity, readability and discovery.

Instead of a function where we pass the Scheme to Authenticate() and get a result we now must construct a context object that will be mutated by the Authenticate method.  

This approach forces me to ask a ton of questions about API:
- How should the context object be constructed? Is the ctor parameter enough? Do I also need to set properties on it? Does it need to be passed to another method first for other mutations?
- How do I determine failure?  Will the method throw if Authentication fails?  What property on the context reflects success?
- What properties will success update on the context object?  Which properties are inputs, which are outputs?  Is that context object thread safe?  What happens if I pass it elsewhere and some other code is also passing it to an another Authenticate method?

Functions/Methods have signatures for a reason.  They have input parameters and they have return values.  This allows to reason about the effects and non-effects of calling them.  This use of context objects completely eliminates these benefits.  Why are we doing it?

HaoK commented:
Just to be clear, there still is the following API (which now directly give syou the principal):

```
ClaimsPrincipal Authenticate(string authenticationScheme)
```

Its only if you are using the less commonly accessed AuthenticationProperties/Description where you have to use the context

darrelmiller commented:
@HaoK But that's syntactic sugar over the primary API which uses the context object, right?

HaoK commented:
Depends on how far you want to take that line of argument, the whole AuthenticationManager API is 'sugar' on top of the HttpFeature which has always looked this way (AuthenticateContext)...  The most commonly used API should be the new simpler ClaimsPrincipal Authenticate.  We removed the intermediate API which returned a tuple of 3 things from the Context.  The intent is to guide people to the new simplier API.

darrelmiller commented:
I agree that the simpler ClaimsPrincipal Authenticate is better than the API that returned AuthenticationResult.  And if the objective is to guide people towards the simpler API for the majority of cases then I would agree that is a good thing.
I'm just frustrated by context objects throughout the framework being the primary interaction method.  Glad to hear that's not the case here.

davidfowl commented on src/Microsoft.AspNet.Http/Authentication/DefaultAuthenticationManager.cs:
```diff
@@ -117,6 +95,18 @@ public override void Challenge(string authenticationScheme, AuthenticationProper
         }
 
+        // You are not allowed access
+        public void Forbidden(string authenticationScheme, AuthenticationProperties properties)
```
Deny?

HaoK replied:
Are we changing the ChallengeBehavior enum to match the naming as well?

davidfowl replied:
Forbid?

PinpointTownes commented:
> Functions/Methods have signatures for a reason. They have input parameters and they have return values. This allows to reason about the effects and non-effects of calling them. This use of context objects completely eliminates these benefits. Why are we doing it?

I couldn't agree more with you, sir :+1:

HaoK commented:
Renamed Forbidden => Forbid and added overloads to abstract base class to make it friendlier (optional AuthenticationProperties, and exposed the raw abstract Challenge overload which takes ChallengeBehavior which will make @lodejard happy)

HaoK commented:
One thing I couldn't figure out how to do yet is remove the 401 setting inside of Challenge, @lodejard was adamant that this behavior belongs in the IAuthenticationHandler instead, I haven't quite gotten that to work yet.

HaoK commented:
Per discussion with @Tratcher made AuthManager apis async to try and eliminate duplicate code paths in AuthHandlers.

PinpointTownes commented on src/Microsoft.AspNet.Http.Abstractions/Authentication/AuthenticationManager.cs:
```diff
@@ -11,41 +14,93 @@ public abstract class AuthenticationManager
+        private static TResult RunSync<TResult>(Func<Task<TResult>> func)
+        {
+            // REVIEW: Do we need to flow culture info still?
+            return _myTaskFactory.StartNew(() => func()).Unwrap().GetAwaiter().GetResult();
```
OMFG, what's that? :scream:

It's bad, bad, bad, horribly bad: you're offloading an **async** delegate (`() => func()`) to the thread pool before unwrapping it and then blocking on this new task :smile: 

This clearly creates unnecessary pressure on the thread pool, specially since you use `StartNew` on an `Task`-returning delegate, which is supposed to be non-blocking.

HaoK replied:
Yeah well this was the only safe way to do this historically in asp.net (calling async methods synch). Other less nasty ways would deadlock in some situations. I'm not sure this is still needed in the new world.

Tratcher replied:
I'd be really surprised if this was still necessary now that we no longer have the System.Web call context.

HaoK replied:
Ok cool, so its definitely not as gross with out this...

@davidfowl What do you think about this change?  (Making everything async only for auth, and only having a sync public wrapper in the manager?

HaoK commented:
After discussing with @lodejard @Tratcher @Eilon concluded that we are going with async only, removing prototype tag, these should be the changes needed for security.

PinpointTownes commented on src/Microsoft.AspNet.Owin/OwinEnvironment.cs:
```diff
@@ -61,7 +61,7 @@ public OwinEnvironment(HttpContext context)
                 { OwinConstants.ResponseHeaders, new FeatureMap<IHttpResponseFeature>(feature => feature.Headers, (feature, value) => feature.Headers = (IDictionary<string, string[]>)value) },
                 { OwinConstants.ResponseBody, new FeatureMap<IHttpResponseFeature>(feature => feature.Body, () => Stream.Null, (feature, value) => feature.Body = (Stream)value) },
-                { OwinConstants.CommonKeys.OnSendingHeaders, new FeatureMap<IHttpResponseFeature>(feature => new Action<Action<object>, object>(feature.OnSendingHeaders)) },
+                { OwinConstants.CommonKeys.OnSendingHeaders, new FeatureMap<IHttpResponseFeature>(feature => new Action<Func<object, Task>, object>(feature.OnResponseStarting)) },
```
I'm not sure you can change the signature of `server.OnSendingHeaders` without breaking tons of OWIN middleware that expect a `Action<Action<object>, object>`.

http://owin.org/spec/spec/CommonKeys.html

PinpointTownes commented on src/Microsoft.AspNet.Http.Abstractions/HttpResponse.cs:
```diff
@@ -22,11 +23,11 @@ public abstract class HttpResponse
+        public abstract bool HasStarted { get; }
 
-        public abstract void OnSendingHeaders(Action<object> callback, object state);
+        public abstract void OnResponseStarting(Func<object, Task> callback, object state);
```
Yeah, does that mean OSH callbacks will be async only? :smile: 
Is it worth re-opening the ticket in aspnet/Security then?

Tratcher commented on src/Microsoft.AspNet.Http.Abstractions/Authentication/AuthenticationManager.cs:
```diff
@@ -11,41 +13,66 @@ public abstract class AuthenticationManager
 
-        public abstract void Challenge(string authenticationScheme, AuthenticationProperties properties);
+        // Leave it up to authentication handler to do the right thing for the challenge
+        public Task ChallengeAsync([NotNull] string authenticationScheme, AuthenticationProperties properties)
```
You're using `[NotNull]` here but then passing in null from the overloads. This can only end well...

HaoK replied:
Whoops I had a bad feeling when I added them yesterday while rebasing to dev. Now I know why!  Good catch :)

HaoK replied:
Updated PR to use [NotNull] everywhere, Challenge sugar methods now pass string.Empty since we were already doing string.IsNullOrWhitespace checks, this shouldn't affect anything.  Also nuked the overloads of SignIn/SignOut that didn't require a scheme, this was on the todo list from earlier API reviews with Lou

PinpointTownes replied:
IMHO, using `string.Empty` is a bit ugly. And removing `SignIn/SignOut` overloads that don't take a specific scheme while keeping the `Challenge` method that takes no parameter is a bit weird, no?

HaoK replied:
Naked Challenge is required for naked Authorize which is pretty important for MVC.

SignIn/SignOut with no arguments where any automatic middleware picks that up felt like a pretty unnecessary feature.  

We are going to try and make it explicit so authenticationScheme is more understood since its always there...

PinpointTownes replied:
Well, `SignIn/SignOut with no arguments` are a great way for the underlying application to keep the separation of concerns clear. For instance, controllers shouldn't have to know which schemes have been configured in Startup.cs request to globally sign in or sign out the user.

HaoK replied:
That's what identity is for.  The feedback has been pretty consistent that AutomaticAuthentication is confusing.  We need it for naked challenge and the automatic cookie/bearer flow.  But for everything else we want to dial that back is our current thinking.

Tratcher commented on src/Microsoft.AspNet.Http.Abstractions/Authentication/AuthenticationManager.cs:
```diff
@@ -11,41 +13,66 @@ public abstract class AuthenticationManager
+        }
 
-        public void SignIn(string authenticationScheme, ClaimsPrincipal principal)
+        public Task SignInAsync([NotNull] string authenticationScheme, ClaimsPrincipal principal)
```
nit, organization: Move this down to be with the rest of SignIn

Tratcher commented on src/Microsoft.AspNet.Http.Abstractions/Authentication/AuthenticationManager.cs:
```diff
@@ -11,41 +13,66 @@ public abstract class AuthenticationManager
+        }
 
-        public abstract void SignOut(string authenticationScheme, AuthenticationProperties properties);
+        public abstract Task SignOutAsync([NotNull] string authenticationScheme, AuthenticationProperties properties);
```
Out of curiosity, what are you expecting [NotNull] to do on an abstract method? I found EF doing the same thing recently.

HaoK replied:
I think its so tools like resharper can warn you

davidfowl replied:
docs yep

Tratcher commented on src/Microsoft.AspNet.Http.Abstractions/HttpResponse.cs:
```diff
@@ -24,9 +25,9 @@ public abstract class HttpResponse
         public abstract bool HasStarted { get; }
 
-        public abstract void OnResponseStarting(Action<object> callback, object state);
+        public abstract void OnResponseStarting(Func<object, Task> callback, object state);
```
Do we want sync sugar for these? And while we're at it we could have sugar that doesn't require state.

davidfowl replied:
Documentation

HaoK replied:
I filed https://github.com/aspnet/HttpAbstractions/issues/333 to track.

Lou said no sync sugar anywhere as part of agreeing to full async.  I guess we are strongly encouraging async this way. I'll add a stateless sugar overload now

Tratcher replied:
Lou said no sync sugar that calls async and Wait. Here you'd just wrap the Action in a Func and return a completed Task.

That said, does it confuse the intellisense for anonymous delegates if you have overloads that differ only by return value?

HaoK replied:
Fair enough, updated with all the new overloads

PinpointTownes commented on src/Microsoft.AspNet.Owin/OwinEnvironment.cs:
```diff
@@ -61,7 +61,7 @@ public OwinEnvironment(HttpContext context)
                 { OwinConstants.ResponseHeaders, new FeatureMap<IHttpResponseFeature>(feature => feature.Headers, (feature, value) => feature.Headers = (IDictionary<string, string[]>)value) },
                 { OwinConstants.ResponseBody, new FeatureMap<IHttpResponseFeature>(feature => feature.Body, () => Stream.Null, (feature, value) => feature.Body = (Stream)value) },
-                { OwinConstants.CommonKeys.OnSendingHeaders, new FeatureMap<IHttpResponseFeature>(feature => new Action<Action<object>, object>(feature.OnResponseStarting)) },
+                { OwinConstants.CommonKeys.OnSendingHeaders, new FeatureMap<IHttpResponseFeature>(feature => new Action<Func<object, Task>, object>(feature.OnResponseStarting)) },
```
So, what did you decide for OWIN?

HaoK replied:
@Tratcher mentioned something about adapting this, I haven't had a chance to take a look at this yet.  I'm guessing we don't have any test coverage of this?

Tratcher replied:
Yeah, this will require some adaptation. OWIN middleware on ASP.NET servers will be easy because we can wrap an `Action<object>` in a `Func<object, Task>`. However, running ASP.NET components on OWIN servers will be a little awkward because we have to wrap a  `Func<object, Task>` in a `Action<object>` and block.

PinpointTownes replied:
Nice, that's exactly what I had in mind :smile:

HaoK commented on src/Microsoft.AspNet.Http.Abstractions/HttpResponse.cs:
```diff
@@ -24,9 +25,14 @@ public abstract class HttpResponse
+        {
+            OnResponseStarting(callback, state: null);
+        }
+
```
@Tratcher is this sweet enough?

HaoK replied:
Perhaps we should add one for OnResponseCompleted too?

Tratcher replied:
How about `OnResponseStarting(Func<Task> callback) `?

Sure, add the same for OnResponseCompleted too.

PinpointTownes replied:
We definitely need more C# 6 goodness: 

```csharp
public void OnResponseStarting(Func<Task> callback) => OnResponseStarting(state => callback(state), state: null);
```

Tratcher commented on src/Microsoft.AspNet.Http.Abstractions/HttpResponse.cs:
```diff
@@ -24,14 +25,33 @@ public abstract class HttpResponse
-        }
+        public virtual void OnResponseStarting(Action callback) => OnResponseStarting(s => callback(), state: null);
+
+        public virtual void OnResponseStarting(Action<object> callback, object state) =>
```
Does the intellisense get confused with anonymous delegates between Action and Func?

HaoK replied:
Yeah looks like Func wins, so it doesn't let you provide a naked Action, says wrong return type for the void Action oh well

Tratcher commented on src/Microsoft.AspNet.Http.Abstractions/HttpResponse.cs:
```diff
@@ -24,14 +25,33 @@ public abstract class HttpResponse
-        {
-            Redirect(location, permanent: false);
-        }
+        public virtual void OnResponseStarting(Action callback) => OnResponseStarting(s => callback(), state: null);
```
Extra closure. You can pass the callback into a static Func as state.

HaoK replied:
I'm not sure I follow, aren't all of these sugar methods supposed to be delegating to abstract OnResponseStarting?

PinpointTownes replied:
I guess @Tratcher suggests something like that:

```csharp
public void OnResponseStarting(Func<Task> callback) => OnResponseStarting(cb => ((Func<Task>) cb)(), state: callback);
```

Tratcher replied:
Yes, but as a static method/delegate to prevent allocations for every call.

HaoK replied:
Not following I think still, can you outline the static method that you are envisioning?

PinpointTownes replied:
```csharp
private static readonly Func<object, Task> _onResponseStartingDelegate = callback => ((Func<Task>) callback)();

public void OnResponseStarting(Func<Task> callback) => OnResponseStarting(_onResponseStartingDelegate, state: callback);
```

HaoK replied:
Ah got it, so are we still not able to use this overload in Security AuthHandler after the change to a static delegate?

HaoK replied:
Updated nuking the Action overloads, and switching both sugar methods to use the same static delegate

HaoK commented:
Added a new `DisposeOnCompleted(IDisposable)` since MVC was using this all over.  Seemed generally useful.  @Tratcher what do you think about shortening the name of the OnResponseXyz methods to simply OnXyz()?  They already hang off of httpResponse, it seems a bit redundant to say response.OnResponseStarted/Completed.  

response.OnStarted/OnCompleted seems just as clear...  I'm touching all of the signatures already, so if we want to rename these, its a free opportunity

Tratcher commented:
Yes, OnStarted/Completed is better.

HaoK commented:
Updated with OnStarted/OnCompleted/OnCompetedDispose

Tratcher commented on src/Microsoft.AspNet.Http.Abstractions/HttpResponse.cs:
```diff
@@ -24,14 +32,17 @@ public abstract class HttpResponse
-        }
+        public abstract void OnCompleted(Func<object, Task> callback, object state);
+
+        public virtual void OnCompletedDispose(IDisposable disposable) => OnStarting(_disposeDelegate, disposable);
```
This should be OnCompleted, not OnStarting.

HaoK replied:
That would explain the functional failures I was seeing :)

Tratcher commented:
Yes, that looks right.

Tratcher commented on src/Microsoft.AspNet.Http.Abstractions/HttpResponse.cs:
```diff
@@ -24,14 +32,17 @@ public abstract class HttpResponse
-        }
+        public abstract void OnCompleted(Func<object, Task> callback, object state);
+
+        public virtual void OnCompletedDispose(IDisposable disposable) => OnCompleted(_disposeDelegate, disposable);
```
`[NotNull]` for everything but `state`.

Tratcher commented:
:shipit:

HaoK commented:
5fe8037281bb826e0708abdcdafbc76571dc21f5
